In [53]:
import pandas as pd

In [54]:
data = pd.read_csv('data/training.csv')

In [55]:
data.isna().sum().sum()

np.int64(0)

U datasetu nema null vrednosti.

In [56]:
data.select_dtypes(include='str')

,PIDN,Depth
0,XNhoFZW5,Topsoil
1,9XNspFTd,Subsoil
2,WDId41qG,Topsoil
3,JrrJf1mN,Subsoil
4,ZoIitegA,Topsoil
...,...,...
1152,bdcNNrbi,Topsoil
1153,6HBVKZwh,Subsoil
1154,5eLY5nw7,Topsoil
1155,gsSGXhX6,Subsoil


Izbacujemo PDIN koji predstavlja identifikator uzorka

In [57]:
data.drop('PIDN', inplace=True, axis=1)

Pretvaramo Depth obelezje u numericke vrednosti

In [58]:
data['Depth'] = data['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})

## Provera korelacija

In [59]:
import numpy as np

cor_mat = data.corr()
cor_mat_ut = cor_mat.where(np.triu(np.ones(cor_mat.shape), k=1).astype(np.bool))
cols = cor_mat_ut.columns
correlated_columns = []
for column in cols:
    if any(cor_mat_ut[column] > 0.95):
        correlated_columns.append(column)
correlated_columns

['m7496.04',
 'm7494.11',
 'm7492.18',
 'm7490.25',
 'm7488.32',
 'm7486.39',
 'm7484.46',
 'm7482.54',
 'm7480.61',
 'm7478.68',
 'm7476.75',
 'm7474.82',
 'm7472.89',
 'm7470.97',
 'm7469.04',
 'm7467.11',
 'm7465.18',
 'm7463.25',
 'm7461.32',
 'm7459.39',
 'm7457.47',
 'm7455.54',
 'm7453.61',
 'm7451.68',
 'm7449.75',
 'm7447.82',
 'm7445.89',
 'm7443.97',
 'm7442.04',
 'm7440.11',
 'm7438.18',
 'm7436.25',
 'm7434.32',
 'm7432.4',
 'm7430.47',
 'm7428.54',
 'm7426.61',
 'm7424.68',
 'm7422.75',
 'm7420.82',
 'm7418.9',
 'm7416.97',
 'm7415.04',
 'm7413.11',
 'm7411.18',
 'm7409.25',
 'm7407.33',
 'm7405.4',
 'm7403.47',
 'm7401.54',
 'm7399.61',
 'm7397.68',
 'm7395.75',
 'm7393.83',
 'm7391.9',
 'm7389.97',
 'm7388.04',
 'm7386.11',
 'm7384.18',
 'm7382.25',
 'm7380.33',
 'm7378.4',
 'm7376.47',
 'm7374.54',
 'm7372.61',
 'm7370.68',
 'm7368.76',
 'm7366.83',
 'm7364.9',
 'm7362.97',
 'm7361.04',
 'm7359.11',
 'm7357.18',
 'm7355.26',
 'm7353.33',
 'm7351.4',
 'm7349.47',
 'm734

In [60]:
data_bez_korelacija =  data.drop(correlated_columns, axis=1)

In [61]:
data_bez_korelacija

,m7497.96,BSAN,BSAV,CTI,ELEV,EVI,LSTD,LSTN,REF2,REF3,RELI,TMAP,TMFI,Depth,Ca,P,pH,SOC,Sand
0,0.302553,-0.630435,-0.783875,-0.364146,1.165479,1.062682,-0.716713,-0.090016,-0.537106,-0.722567,1.687734,0.190708,0.056843,0,-0.295749,-0.041336,-1.129366,0.353258,1.269748
1,0.270192,-0.630435,-0.783875,-0.364146,1.165479,1.062682,-0.716713,-0.090016,-0.537106,-0.722567,1.687734,0.190708,0.056843,1,-0.387442,-0.231552,-1.531538,-0.264023,1.692209
2,0.317433,-0.753623,-0.929451,-0.633972,1.544098,1.156705,-1.282552,-0.088336,-0.631725,-0.832298,1.806660,0.190708,0.056843,0,-0.248601,-0.224635,-0.259551,0.064152,2.091835
3,0.261116,-0.753623,-0.929451,-0.633972,1.544098,1.156705,-1.282552,-0.088336,-0.631725,-0.832298,1.806660,0.190708,0.056843,1,-0.332195,-0.318014,-0.577548,-0.318719,2.118477
4,0.260038,-0.688406,-0.884658,-0.583576,1.276837,1.191691,-1.206971,0.011420,-0.528757,-0.795031,0.430513,0.190708,0.056843,0,-0.438350,-0.010210,-0.699135,-0.310905,2.164148
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1152,0.207530,-0.724638,-0.772676,0.038610,1.501782,0.306851,-0.511141,-0.481023,-0.940631,-0.966874,1.772681,1.539208,1.618022,0,-0.319468,-0.307638,-0.895545,3.056795,-0.971962
1153,0.182376,-0.724638,-0.772676,0.038610,1.501782,0.306851,-0.511141,-0.481023,-0.940631,-0.966874,1.772681,1.539208,1.618022,1,-0.309633,-0.380266,-0.727194,2.838012,-0.983380
1154,0.146829,-0.695652,-0.795073,-0.639966,1.287973,0.659621,-0.549317,-0.646528,-0.515770,-0.913043,2.944954,1.539208,1.618022,0,-0.446449,-0.380266,-0.409197,4.369495,-0.404874
1155,0.091910,-0.695652,-0.795073,-0.639966,1.287973,0.659621,-0.549317,-0.646528,-0.515770,-0.913043,2.944954,1.539208,1.618022,1,-0.493019,-0.418309,-0.194081,2.533278,-0.568530


racunanje korelacije mi je ubilo operativni sistem

## Outlier-i

In [62]:
from scipy.stats import normaltest

norm = []
columns = data.select_dtypes(include='number').columns.tolist()
for column in columns:
    p = normaltest(data[column])
    if p.pvalue < 0.05:
        norm.append(False)
    else:
        norm.append(True)

print(pd.DataFrame(norm).sum(), data.shape)

0    418
dtype: int64 (1157, 3599)


Od 3599 kolona svega 418 ima normalnu raspodelu, pa ne mozemo koristiti z_score za uklanjanje outlier-a

In [63]:
data_numeric = data.select_dtypes(include='number')

data_bez_outliera = data

q1 = data_numeric.quantile(0.25)
q3 = data_numeric.quantile(0.75)
iqr = q3 - q1

columns = data_numeric.columns
for column in columns:
    if (column == 'Ca') | (column == 'P') | (column == 'pH') | (column == 'SOC') | (column == 'Sand'):
        continue
    data_bez_outliera = data_bez_outliera[(data_bez_outliera[column] >= q1[column] - 1.5 * iqr[column]) & (data_bez_outliera[column] <= q3[column] + 1.5 * iqr[column])]

In [64]:
data_bez_outliera.shape

(715, 3599)

Izbacivanjem outliera izgubili smo preveliku kolicinu podataka, pa cemo ih clippovati

In [65]:
data_numeric = data.select_dtypes(include='number')

data_clipped = data

q1 = data_numeric.quantile(0.25)
q3 = data_numeric.quantile(0.75)
iqr = q3 - q1

columns = data_numeric.columns
for column in columns:
    if (column == 'Ca') | (column == 'P') | (column == 'pH') | (column == 'SOC') | (column == 'Sand'):
        continue
    min = q1[column] - 1.5 * iqr[column]
    max = q3[column] + 1.5 * iqr[column]
    data_clipped[column] = data_clipped[column].clip(upper=max, lower=min)

## Modeli

Linearna regresija

In [66]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

model_lin_reg = LinearRegression(n_jobs=-1)
pipeline_lin_reg = Pipeline([
    ('scaler', StandardScaler()),
    ('model_lin_reg', model_lin_reg),
])
grid_params = {
    'scaler' : [None, StandardScaler()],
    'model_lin_reg__fit_intercept' : [False, True],
}

grid_lin_reg = GridSearchCV(pipeline_lin_reg, grid_params, n_jobs=-1, cv=5, scoring='neg_mean_squared_error')

Klasifikaciono stablo

In [67]:
from sklearn.tree import DecisionTreeRegressor

model_tree = DecisionTreeRegressor(random_state=7)
pipeline_tree = Pipeline([
    #('scaler', StandardScaler()),
    ('model_tree', model_tree),
])

parameters_tree = {
    "model_tree__ccp_alpha" : [0.0, 0.05, 0.0025],
    "model_tree__max_depth" : [2, 5, 7, 9, 11],
    "model_tree__criterion" : ['squared_error', 'friedman_mse', 'absolute_error'],
}

grid_tree = GridSearchCV(pipeline_tree, parameters_tree, n_jobs=-1, cv=5, scoring='neg_mean_squared_error')

## Sa outlierima

Izdvajamo izlazna obelezja

In [68]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import root_mean_squared_error

def the_funkcija(grid, trening_set, validacija_set, output_name):
    # izdvajanje izlaznih obelezja
    y_Ca = trening_set['Ca']
    y_P  = trening_set['P']
    y_pH = trening_set['pH']
    y_SOC = trening_set['SOC']
    y_Sand = trening_set['Sand']
    X = trening_set.drop(['Ca', 'P', 'pH', 'SOC', 'Sand'], axis=1)
    # train test split
    X_train_Ca, X_test_Ca, y_train_Ca, y_test_Ca = train_test_split(X, y_Ca, test_size=0.2, random_state=7)
    X_train_P, X_test_P, y_train_P, y_test_P = train_test_split(X, y_P, test_size=0.2, random_state=7)
    X_train_pH, X_test_pH, y_train_pH, y_test_pH = train_test_split(X, y_pH, test_size=0.2, random_state=7)
    X_train_SOC, X_test_SOC, y_train_SOC, y_test_SOC = train_test_split(X, y_SOC, test_size=0.2, random_state=7)
    X_train_Sand, X_test_Sand, y_train_Sand, y_test_Sand = train_test_split(X, y_Sand, test_size=0.2, random_state=7)
    # krosvalidacija
    grid_Ca = grid.fit(X_train_Ca, y_train_Ca)
    best_model_Ca = grid_Ca.best_estimator_
    best_params_Ca = grid_Ca.best_params_

    grid_P = grid.fit(X_train_P, y_train_P)
    best_model_P = grid_P.best_estimator_
    best_params_P = grid_P.best_params_

    grid_pH = grid.fit(X_train_pH, y_train_pH)
    best_model_pH = grid_pH.best_estimator_
    best_params_pH = grid_pH.best_params_

    grid_SOC = grid.fit(X_train_SOC, y_train_SOC)
    best_model_SOC = grid_SOC.best_estimator_
    best_params_SOC = grid_SOC.best_params_

    grid_Sand = grid.fit(X_train_Sand, y_train_Sand)
    best_model_Sand = grid_Sand.best_estimator_
    best_params_Sand = grid_Sand.best_params_

    # kalkulacija rmse
    y_RMSE_Ca = root_mean_squared_error(best_model_Ca.predict(X_test_Ca), y_test_Ca)
    y_RMSE_P = root_mean_squared_error(best_model_P.predict(X_test_P), y_test_P)
    y_RMSE_pH = root_mean_squared_error(best_model_pH.predict(X_test_pH), y_test_pH)
    y_RMSE_SOC = root_mean_squared_error(best_model_SOC.predict(X_test_SOC), y_test_SOC)
    y_RMSE_Sand = root_mean_squared_error(best_model_Sand.predict(X_test_Sand), y_test_Sand)

    # validacija
    prediction = pd.DataFrame()
    prediction['PIDN'] = validacija_set['PIDN']
    validacija_set.drop('PIDN', inplace=True, axis=1)

    prediction['Ca'] = pd.DataFrame(best_model_Ca.set_params(**best_params_Ca).fit(X, y_Ca).predict(validacija_set))
    prediction['P'] = pd.DataFrame(best_model_P.set_params(**best_params_P).fit(X, y_P).predict(validacija_set))
    prediction['pH'] = pd.DataFrame(best_model_pH.set_params(**best_params_pH).fit(X, y_pH).predict(validacija_set))
    prediction['SOC'] = pd.DataFrame(best_model_SOC.set_params(**best_params_SOC).fit(X, y_SOC).predict(validacija_set))
    prediction['Sand'] = pd.DataFrame(best_model_Sand.set_params(**best_params_Sand).fit(X, y_Sand).predict(validacija_set))

    prediction.to_csv("predictions/" + output_name + ".csv", index=False)

    return {
        "Ca" : y_RMSE_Ca,
        "P" : y_RMSE_P,
        "pH": y_RMSE_pH,
        "SOC": y_RMSE_SOC,
        "Sand": y_RMSE_Sand,
        "MCRMSE" : (y_RMSE_Ca + y_RMSE_P + y_RMSE_pH + y_RMSE_SOC + y_RMSE_Sand)/5,
    }

In [69]:
sorted_test = pd.read_csv('data/sorted_test.csv')
test_lin_reg = sorted_test.copy()
test_lin_reg['Depth'] = test_lin_reg['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})

In [70]:
print(the_funkcija(grid_lin_reg, data, test_lin_reg, "lin_reg"))

{'Ca': 0.5049645752975458, 'P': 1.6576700943521183, 'pH': 0.7543211407602292, 'SOC': 0.7051158437573091, 'Sand': 0.7294245874734764, 'MCRMSE': 0.8702992483281358}


In [71]:
test_tree = sorted_test.copy()
test_tree['Depth'] = test_tree['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})

In [72]:
print(the_funkcija(grid_tree, data, test_tree, "tree"))

{'Ca': 0.5648514501553454, 'P': 1.0588995128583363, 'pH': 0.6745977967977185, 'SOC': 0.5302018230051582, 'Sand': 0.5241819155477588, 'MCRMSE': 0.6705464996728635}


## Bez outliera

In [73]:
test_lin_reg_c = sorted_test.copy()
test_lin_reg_c['Depth'] = test_lin_reg_c['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})

In [74]:
print(the_funkcija(grid_lin_reg, data_clipped, test_lin_reg_c, "lin_reg_clipped"))

{'Ca': 0.5049645752975458, 'P': 1.6576700943521183, 'pH': 0.7543211407602292, 'SOC': 0.7051158437573091, 'Sand': 0.7294245874734764, 'MCRMSE': 0.8702992483281358}


In [75]:
test_tree_clipped = sorted_test.copy()
test_tree_clipped['Depth'] = test_tree_clipped['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})

In [76]:
print(the_funkcija(grid_tree, data_clipped, test_tree_clipped, "tree_clipped"))

{'Ca': 0.5648514501553454, 'P': 1.0588995128583363, 'pH': 0.6745977967977185, 'SOC': 0.5302018230051582, 'Sand': 0.5241819155477588, 'MCRMSE': 0.6705464996728635}


Zakljucujemo da linearna regresija trenirana na clippovanim podacima radi isto

## Bez korelisanih kolona

In [77]:
test_lin_reg_bk = sorted_test.copy()
test_lin_reg_bk['Depth'] = test_lin_reg_bk['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})
test_lin_reg_bk = test_lin_reg_bk.drop(correlated_columns, axis=1)

In [78]:
print(the_funkcija(grid_lin_reg, data_bez_korelacija, test_lin_reg_bk, "lin_reg_bez_korelacija"))

{'Ca': 0.8329978812820192, 'P': 0.6988197190977719, 'pH': 0.6801143967048152, 'SOC': 0.8315303328421029, 'Sand': 0.7703700255047135, 'MCRMSE': 0.7627664710862845}


In [79]:
test_tree_bk = sorted_test.copy()
test_tree_bk['Depth'] = test_tree_bk['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})
test_tree_bk = test_tree_bk.drop(correlated_columns, axis=1)

In [80]:
print(the_funkcija(grid_tree, data_bez_korelacija, test_tree_bk, "tree_bez_korelacija"))

{'Ca': 0.5108288448240199, 'P': 0.7250483854005842, 'pH': 0.6163802595768648, 'SOC': 0.722614972588117, 'Sand': 0.5259174484953628, 'MCRMSE': 0.6201579821769898}


## Polynomial features

broj polinoma je prevelik pre izbacivanja korelisanih kolona, pa samo ovde koristimo polynomial features

ostavili smo samo drugi stepen polinoma jer visi stepeni daju isti rezulatat

In [81]:
from sklearn.preprocessing import PolynomialFeatures

pipeline_lin_reg_polynomial = Pipeline([
    ('scaler', StandardScaler()),
    ('feature_generator', PolynomialFeatures()),
    ('model_lin_reg', model_lin_reg),
])
grid_params = {
    'scaler' : [None, StandardScaler()],
    'feature_generator' : [None, PolynomialFeatures(degree=2)],
    'model_lin_reg__fit_intercept' : [False, True],
}

grid_lin_reg_polynomial = GridSearchCV(pipeline_lin_reg_polynomial, grid_params, n_jobs=-1, cv=5, scoring='neg_mean_squared_error')

In [82]:
test_lin_reg_bk_pol = sorted_test.copy()
test_lin_reg_bk_pol['Depth'] = test_lin_reg_bk_pol['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})
test_lin_reg_bk_pol = test_lin_reg_bk_pol.drop(correlated_columns, axis=1)

In [83]:
print(the_funkcija(grid_lin_reg_polynomial, data_bez_korelacija, test_lin_reg_bk_pol, "lin_reg_bez_korelacija_polynomial"))

{'Ca': 0.6420833025861913, 'P': 0.6988197190977719, 'pH': 0.6189733116872884, 'SOC': 0.766177253697109, 'Sand': 0.6706059297434439, 'MCRMSE': 0.6793319033623609}


In [84]:

pipeline_tree_polynomial = Pipeline([
    #('scaler', StandardScaler()),
    ('feature_generator', 'passthrough'),
    ('model_tree', model_tree),
])

parameters_tree = {
    "model_tree__ccp_alpha" : [0.0, 0.05, 0.0025],
    "model_tree__max_depth" : [2, 5, 7, 9, 11],
    "model_tree__criterion" : ['squared_error', 'friedman_mse', 'absolute_error'],
    "feature_generator" : ['passthrough', PolynomialFeatures(degree=2)]
}

grid_tree_polynomial = GridSearchCV(pipeline_tree_polynomial, parameters_tree, n_jobs=-1, cv=5, scoring='neg_mean_squared_error')

In [85]:
test_tree_bk_pol = sorted_test.copy()
test_tree_bk_pol['Depth'] = test_tree_bk_pol['Depth'].replace({'Topsoil' : 0 , 'Subsoil' : 1})
test_tree_bk_pol = test_tree_bk_pol.drop(correlated_columns, axis=1)

In [86]:
print(the_funkcija(grid_tree_polynomial, data_bez_korelacija, test_tree_bk_pol, "tree_bez_korelacija_polynomial"))

{'Ca': 0.5108288448240199, 'P': 0.7250483854005842, 'pH': 0.6163802595768648, 'SOC': 0.722614972588117, 'Sand': 0.5259174484953628, 'MCRMSE': 0.6201579821769898}
